File: day_02_valid_anagram/solution.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        char_counts = {}
        for char in s:
            char_counts[char] = char_counts.get(char, 0) + 1
            
        for char in t:
            if char not in char_counts:
                return False
            char_counts[char] -= 1
            if char_counts[char] == 0:
                del char_counts[char]
                
        return len(char_counts) == 0

# Simple Verification Tests
if __name__ == "__main__":
    sol = Solution()
    
    # Test Case 1
    assert sol.isAnagram("anagram", "nagaram") is True
    
    # Test Case 2
    assert sol.isAnagram("rat", "car") is False
    
    # Test Case 3
    assert sol.isAnagram("a", "ab") is False
    
    print("Day 2 (Valid Anagram): All tests passed successfully! 🎉")

File: day_02_valid_anagram/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <cassert>

class Solution {
public:
    bool isAnagram(std::string s, std::string t) {
        if (s.length() != t.length()) {
            return false;
        }
        
        std::unordered_map<char, int> char_counts;
        for (char c : s) {
            char_counts[c]++;
        }
        
        for (char c : t) {
            if (char_counts.find(c) == char_counts.end()) {
                return false;
            }
            char_counts[c]--;
            if (char_counts[c] == 0) {
                char_counts.erase(c);
            }
        }
        
        return char_counts.empty();
    }
};

int main() {
    Solution sol;
    
    // Test Case 1
    assert(sol.isAnagram("anagram", "nagaram") == true);
    
    // Test Case 2
    assert(sol.isAnagram("rat", "car") == false);
    
    // Test Case 3
    assert(sol.isAnagram("a", "ab") == false);
    
    std::cout << "Day 2 (Valid Anagram) C++: All tests passed successfully! 🎉" << std::endl;
    return 0;
}
